core/modules/loader/ServerTcpBase.cc

```diff
@@ -0,0 +1,618 @@
+        return;
+    }
+    LOGS(_log, LOG_LVL_INFO, " _handleShiftToRight bytes=" << bytesInMsg << " buf=" << _buf.dumpStr(false));
+    boost::asio::async_read(_socket, boost::asio::buffer(_buf.getWriteCursor(), bytesInMsg),
+            boost::asio::transfer_at_least(bytesInMsg),
+            boost::bind(
+                    &TcpBaseConnection::_handleShiftToRight1,
+                    shared_from_this(),
+                    boost::asio::placeholders::error,
+                    boost::asio::placeholders::bytes_transferred
+            )
+    );
+}
+
+
+void TcpBaseConnection::_handleShiftToRight1(boost::system::error_code const& ec, size_t bytesTrans) {
+    std::string const funcName = "_handleShiftToRight1";
+    if (ec) {
+        LOGS(_log, LOG_LVL_ERROR, funcName << " ec=" << ec);
+        _freeConnect();
+        return;
+    }
+    // Fix the buffer with the information given.
+    _buf.advanceWriteCursor(bytesTrans);
+    LOGS(_log, LOG_LVL_INFO, funcName << " bytes=" << bytesTrans << " _buf" << _buf.dumpStr(false));
+    try {
+        // TODO move as much of this to CentralWorker as possible
+        LOGS(_log, LOG_LVL_INFO, funcName << " parsing bytes=" << bytesTrans <<
+                                 " _buf" << _buf.dumpStr(false));
+        auto protoKeyList = StringElement::protoParse<proto::KeyList>(_buf);
+        if (protoKeyList == nullptr) {
+            throw LoaderMsgErr(ERR_LOC, "protoKeyList nullptr");
+        }
+        // Extract key pairs from the protobuffer
+        int keyCount = protoKeyList->keycount(); // TODO delete keycount from KeyList
+        int sz = protoKeyList->keypair_size();
+        if (keyCount != sz) {
+            LOGS(_log, LOG_LVL_WARN, funcName << " keyCount(" << keyCount << ") != sz(" << sz << ")");
+        }
+        std::vector<CentralWorker::StringKeyPair> keyList;
+        for (int j=0; j < sz; ++j) {
+            proto::KeyInfo const& protoKI = protoKeyList->keypair(j);
+            ChunkSubchunk chSub(protoKI.chunk(), protoKI.subchunk());
+            keyList.push_back(std::make_pair(protoKI.key(), chSub));
+        }
+
+        // Now that the proto buffer was read without error, insert into map and adjust our range.
+        _serverTcpBase->getCentralWorker()->insertKeys(keyList, true);
+
+        // Send the SHIFT_TO_RIGHT_KEYS_RECEIVED response back.
+        _buf.reset();
+        UInt32Element elem(LoaderMsg::SHIFT_TO_RIGHT_RECEIVED);
+        elem.appendToData(_buf);
+        ServerTcpBase::writeData(_socket, _buf);
+        LOGS(_log, LOG_LVL_INFO, funcName << " done dumpKeys " <<
+                                             _serverTcpBase->getCentralWorker()->dumpKeysStr(2));
+    } catch (LoaderMsgErr &msgErr) {
+        LOGS(_log, LOG_LVL_ERROR, msgErr.what());
+        LOGS(_log, LOG_LVL_ERROR, funcName << " key shift failed");
+        _freeConnect();
+        return;
+    }
+    boost::system::error_code ecode;
+    _readKind(ecode, 0); // get next message
```

Comment:
I've already made a comment earlier on calling this method explicitly. Consider refactoring `_readKind`